elastic/eui
# [EuiDataGrid] Use integer values for dimensions derived from integers

## Bug repro steps

- In any browser, set your browser zoom level to anything above 100%, e.g. 110%
- Go to https://eui.elastic.co/v80.0.0/#/tabular-content/data-grid
  - (NOTE: If you zoom in _while_ on this page, you must refresh the page first so the page loads in at 110% zoom)
- Notice that the first example has a scrollbar (when it should not)
- Go to the pagination `Rows per page` in the bottom left corner of the grid, and select 50
- Notice that the datagrid does not dynamically update its height when it should, it stays stuck as scrolling

## Issues affected by this bug

- https://github.com/elastic/kibana/issues/159143
- https://github.com/elastic/kibana/pull/134268 
- and a few others

## Summary of problem

Under certain circumstances, such as when zoom is not 100%, OS level font/zoom levels are not default, child components having a transform actively applied, browser extensions, and probably a whole lot of other scenarios, `getBoundingClientRect().height/width` [will return a decimal number](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Object_Model/Determining_the_dimensions_of_elements). When these close but not exactly the same values are compared with a value derived from a static number in javascript, like with row height from https://github.com/elastic/eui/blob/main/src/components/datagrid/utils/row_heights.ts#L40, pixel math will break. I believe what's happening is this:

`useFinalGridDimensions` will calculate a height/width and apply it to the Grid component, this calculation is done by storing a height/width in hook state, and setting it to a `getBoundingClientRect().height/width` if those measured values do not match the value from `useUnconstrainedHeight`. This normally works, as `useUnconstrainedHeight` is first static 34 when row heights are default, and no scrollbar/header row/footer row has been rendered yet, state is set to 34, and both the EUI components and the grid rerender, and the process happens again. This time with the static 34, and a scrollbar of 16, then once more with a static 34 + 16 + 30 from the header, etc. With integer pixel values, everything is good.

Two things ultimately make the bug happen: the first is the pixel math and `getBoundingClientRect` resulting in decimal values, and the second is that the useEffect that updates the state will run on either height or width changing, and potentially update them both. Sometimes they both are updated, in other cases just the height or width, and when the bug triggers, the effect will typically only update the width, the `unconstrainedHeight` has a value of 50, but for some reason the container is measured as 49.9999997 or something, and so the container height is set to this lower value. From here, no matter what you do, add more rows, change the page etc, the Grid will keep the 49.9997 value as it's height, making the table not usable.

## QA

- In any browser, set your browser zoom level to anything above 100%, e.g. 110%
- Go to https://eui.elastic.co/pr_6895/#/tabular-content/data-grid
  - (NOTE: If you zoom in _while_ on this page, you must refresh the page first so the page loads in at 110% zoom)
- [x] Confirm that the first example **does not** have a scrollbar
- Go to the pagination `Rows per page` in the bottom left corner of the grid, and select 50
- [x] Confirm that the data grid **does** correctly update its height dynamically and adding more rows does not create a scrollbar

### General checklist

- [x] Checked in **mobile** / resizing screens
    - There isn't really an automated test we can add here unfortunately as browser zoom level isn't something we can set w/ Cypress 😢
